### PR TITLE
Add completer for tox 4

### PR DESCRIPTION
**Pull request by samuelallan72**
Add completions for https://tox.wiki/ (based on tox 4.13)

### CODE REVIEW

File: completers/tox_completer/cmd/root.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+func init() {
+	carapace.Gen(rootCmd).Standalone()
+
+	// TODO: support completion of comma separated values
+	rootCmd.Flags().StringS("e", "e", "", "enumerate, comma separated (ALL -> all environments, not set -> use <env_list> from config) (default: <env_list>)")
+
+	carapace.Gen(rootCmd).FlagCompletion(carapace.ActionMap{
+		"e":  tox.ActionEnvironments(),
```
Comment (rsteube) [inline completers/tox_completer/cmd/root.go:27]: there's `.UniqueList(",")` for comma separated values
https://carapace-sh.github.io/carapace/carapace/action/uniqueList.html

**Comment by rsteube** [comment]
Probably best to use the parser here to create the initial command structure: https://github.com/carapace-sh/carapace-bin/tree/master/cmd/carapace-parse

I need to update the docs for it but see the first minute for how it works: https://asciinema.org/a/466859

Store the help output (if available) or https://tox.wiki/en/latest/cli_interface.html of a command (root or subcommand) in a text file.

Prepare it a bit for parsing (something like this):
`-s, --long   shorthand singlespace longhand 2+ space then description`
`-v, --value VALUE  same but accepts a value (single space and some text)`
`--longonly  only longhand`
`-s  only shorthand`

```sh
cat root.txt | carapace-parse -n tox -s "tox root command" > root.go
cat sub.txt | carapace-parse -p sub -p root -s "subcommand of root" > sub.go
cat susbsub2.txt | carapace-parse -s sub2 -p sub -s "subcommand of sub" > sub_sub2.go
```

File: pkg/actions/tools/tox/tox.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import (
+	"strings"
+
+	"github.com/carapace-sh/carapace"
+)
+
+// ActionEnvironments completes tox environments
+func ActionEnvironments() carapace.Action {
+	return carapace.ActionCallback(func(c carapace.Context) carapace.Action {
+		cmd := c.Command("tox", "--listenvs-all")
```
Comment (rsteube) [inline pkg/actions/tools/tox/tox.go:13]: This is fine as it uses `c.Command` which handles context correctly (e.g. setting environment variables).
But there's also [ActionExecCommandE](https://carapace-sh.github.io/carapace/carapace/defaultActions/actionExecCommandE.html) which is intended for custom error handling and might be an option here.

Comment (rsteube) [reply pkg/actions/tools/tox/tox.go:13]: On second thought [ActionExecCommand](https://carapace-sh.github.io/carapace/carapace/defaultActions/actionExecCommand.html) might be sufficient here.

File: completers/tox_completer/cmd/exec.go [hunk 3]
```diff
@@ -14,30 +16,40 @@ var execCmd = &cobra.Command{
...
+
+	carapace.Gen(execCmd).FlagCompletion(carapace.ActionMap{
+		"e":  tox.ActionEnvironments(),
+	})
+
+	carapace.Gen(execCmd).PositionalCompletion(
+		carapace.ActionValues("--"),
+	)
+
+	// TODO: how to get it to complete after `--`?
```
Comment (rsteube) [inline completers/tox_completer/cmd/exec.go:49]: [DashCompletion](https://carapace-sh.github.io/carapace/carapace/gen/dashCompletion.html) and [DashAnyCompletion](https://carapace-sh.github.io/carapace/carapace/gen/dashAnyCompletion.html).
`Context.Args` is reset for these so Actions can be used just as in positional completion.
If you need to continue the positional completion there is [ActionPositional](https://carapace-sh.github.io/carapace/carapace/defaultActions/actionPositional.html).

File: completers/tox_completer/cmd/exec.go [hunk 4]
```diff
@@ -14,30 +16,40 @@ var execCmd = &cobra.Command{
...
-	execCmd.Flags().BoolP("verbose", "v", false, "increase verbosity (default: 2)")
-	execCmd.Flags().Bool("version", false, "show program's and plugins version number and exit")
-	execCmd.Flags().String("workdir", "", "tox working directory (if not specified will be the folder of the config file) (default: None)")
+
+	carapace.Gen(execCmd).FlagCompletion(carapace.ActionMap{
+		"e":  tox.ActionEnvironments(),
+	})
+
+	carapace.Gen(execCmd).PositionalCompletion(
+		carapace.ActionValues("--"),
```
Comment (rsteube) [inline completers/tox_completer/cmd/exec.go:46]: No need to do this yourself. See below.

File: completers/tox_completer/cmd/exec.go [hunk 5]
```diff
@@ -14,30 +16,40 @@ var execCmd = &cobra.Command{
...
 	execCmd.Flags().BoolP("notest", "n", false, "do not run the test commands (default: False)")
-	execCmd.Flags().StringP("pkg-only,", "b", "", "")
-	execCmd.Flags().BoolP("quiet", "q", false, "decrease verbosity (default: 0)")
-	execCmd.Flags().BoolP("recreate", "r", false, "recreate the tox environments (default: False)")
-	execCmd.Flags().String("result-json", "", "write a JSON file with detailed information about all commands and results involved (default: None)")
-	execCmd.Flags().String("root", "", "project root directory (if not specified will be the folder of the config file) (default: None)")
-	execCmd.Flags().String("runner", "", "the tox run engine to use when not explicitly stated in tox env configuration (default: virtualenv)")
-	execCmd.Flags().String("skip-env", "", "exclude all environments selected that match this regular expression (default: )")
+
+	// these two are the same flag; there are two long flags for this
```
Comment (rsteube) [inline completers/tox_completer/cmd/exec.go:32]: That's fine. You can also [mark them mutually exclusive](https://pkg.go.dev/github.com/spf13/cobra@v1.8.1#Command.MarkFlagsMutuallyExclusive) if needed.
So when one is already set the other will not be shown during completion.

File: completers/tox_completer/cmd/exec.go (hunk 3 above)
Comment (samuelallan72) [reply completers/tox_completer/cmd/exec.go:49]: Thanks! :)

File: completers/tox_completer/cmd/exec.go (hunk 5 above)
Comment (samuelallan72) [reply completers/tox_completer/cmd/exec.go:32]: Thanks! :)

**Review by rsteube — APPROVED** [review]
- common.go

Files in `cmd` generally represent a (sub)command.
Moved functions to root.go for consistency since they are small enough.
In another case I've put these in a `common` subfolder (alternative).

- snake_case

Go uses camelCase.

- verbose/quiet

Cobra has `Count` for this kind of flags.

- Common Flags

Seems these are used in every command.
Cobra has `PersistentFlags` which passes them on to subcommands.
It also hides them for subcommands when they are already set on the parent.

- Files / Folder

File completion is explicit in Carapace and needs to be set.
So there is no fallback as shells normally do.
The venv folder name is arbitrary, but it can be an absolute path so folder completion is still beneficial.

- discover, f, m

This is a weird (bad) pattern in some python commands where flags can consume multiple arguments.
I've got basic support for it with `Nargs`.

- StyleF(style.ForKeyword)

No need for this, but highlights positive/negative/arbitrary keywords (like true/false) in colors.

- PreInvoke

This changes the workdir for all actions based on the `workdir` flag value.
(Assuming this is what the flag is meant for)

**Comment by rsteube** [comment]
Made some minor changes. Just open an issue/pr if I botched something.

**Comment by samuelallan72** [comment]
Thanks for your feedback and the merge! :grin:

**Pull request merged**